File: vad/src/main/java/com/konovalov/vad/Validator.java
package com.konovalov.vad;

import com.konovalov.vad.config.FrameSize;
import com.konovalov.vad.config.Model;
import com.konovalov.vad.config.SampleRate;

import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Objects;

final public class Validator {

    private static final LinkedHashMap<SampleRate, LinkedList<FrameSize>> WEB_RTC_VALID_FRAMES =
            new LinkedHashMap<SampleRate, LinkedList<FrameSize>>() {{
                //WEB_RTC supports only 10, 20 or 30 ms frames.
                put(SampleRate.SAMPLE_RATE_8K, new LinkedList<FrameSize>() {{
                    add(FrameSize.FRAME_SIZE_80);
                    add(FrameSize.FRAME_SIZE_160);
                    add(FrameSize.FRAME_SIZE_240);
                }});
                put(SampleRate.SAMPLE_RATE_16K, new LinkedList<FrameSize>() {{
                    add(FrameSize.FRAME_SIZE_160);
                    add(FrameSize.FRAME_SIZE_320);
                    add(FrameSize.FRAME_SIZE_480);
                }});
                put(SampleRate.SAMPLE_RATE_32K, new LinkedList<FrameSize>() {{
                    add(FrameSize.FRAME_SIZE_320);
                    add(FrameSize.FRAME_SIZE_640);
                    add(FrameSize.FRAME_SIZE_960);
                }});
                put(SampleRate.SAMPLE_RATE_48K, new LinkedList<FrameSize>() {{
                    add(FrameSize.FRAME_SIZE_480);
                    add(FrameSize.FRAME_SIZE_960);
                    add(FrameSize.FRAME_SIZE_1440);
                }});
            }};

    private static final LinkedHashMap<SampleRate, LinkedList<FrameSize>> SILERO_VALID_FRAMES =
            new LinkedHashMap<SampleRate, LinkedList<FrameSize>>() {{
                //SILERO supports only 32, 64 or 96 ms frames.
                //Valid frame size: 256 512 768 for 8k; 512 1024 1536 for 16k.
                put(SampleRate.SAMPLE_RATE_8K, new LinkedList<FrameSize>() {{
                    add(FrameSize.FRAME_SIZE_256);
                    add(FrameSize.FRAME_SIZE_512);
                    add(FrameSize.FRAME_SIZE_768);
                }});
                put(SampleRate.SAMPLE_RATE_16K, new LinkedList<FrameSize>() {{
                    add(FrameSize.FRAME_SIZE_512);
                    add(FrameSize.FRAME_SIZE_1024);
                    add(FrameSize.FRAME_SIZE_1536);
                }});
            }};

    public static void validateSampleRate(Model model, SampleRate sampleRate) {
        if (sampleRate == null) {
            throw new NullPointerException("Non-null SampleRate required!");
        }

        if (!getValidSampleRates(model).contains(sampleRate)) {
            throw new IllegalArgumentException(model.name() + " doesn't support Sample Rate: " +
                    sampleRate.getValue() + "!");
        }
    }

    public static void validateFrameSize(Model model, SampleRate sampleRate, FrameSize frameSize) {
        if (frameSize == null) {
            throw new NullPointerException("Non-null FrameSize required!");
        }

        if (!getValidFrameSizes(model, sampleRate).contains(frameSize)) {
            throw new IllegalArgumentException(model.name() + " doesn't support Frame Size: " +
                    frameSize.getValue() + "!");
        }
    }

    public static List<FrameSize> getValidFrameSizes(Model model, SampleRate sampleRate) {
        if (model == null) {
            throw new NullPointerException("Non-null Model required!");
        }

        if (sampleRate == null) {
            throw new NullPointerException("Non-null SampleRate required!");
        }

        switch (model) {
            case WEB_RTC_GMM:
                return new LinkedList<>(Objects.requireNonNull(WEB_RTC_VALID_FRAMES.get(sampleRate)));
            case SILERO_DNN:
                return new LinkedList<>(Objects.requireNonNull(SILERO_VALID_FRAMES.get(sampleRate)));
            default:
                throw new IllegalArgumentException("Model is incorrect!");
        }
    }

    public static List<SampleRate> getValidSampleRates(Model model) {
        if (model == null) {
            throw new NullPointerException("Non-null Model required!");
        }

        switch (model) {
            case WEB_RTC_GMM:
                return new LinkedList<>(WEB_RTC_VALID_FRAMES.keySet());
            case SILERO_DNN:
                return new LinkedList<>(SILERO_VALID_FRAMES.keySet());
            default:
                throw new IllegalArgumentException("Model is incorrect!");
        }
    }
}

File: vad/src/main/java/com/konovalov/vad/Validator.kt
package com.konovalov.vad

import com.konovalov.vad.config.FrameSize
import com.konovalov.vad.config.Model
import com.konovalov.vad.config.SampleRate

/**
 * Created by Georgiy Konovalov on 1/06/2023.
 * <p>
 * Validator responsible for validating sample rates and frame sizes for different VAD models.
 * </p>
 */
object Validator {

    /**
     * <p>
     * Valid sample rates and frame sizes for WebRTC VAD GMM model.
     * </p>
     */
    private val WEB_RTC_VALID_FRAMES: Map<SampleRate, List<FrameSize>> = mapOf(
        SampleRate.SAMPLE_RATE_8K to listOf(
            FrameSize.FRAME_SIZE_80,
            FrameSize.FRAME_SIZE_160,
            FrameSize.FRAME_SIZE_240
        ),
        SampleRate.SAMPLE_RATE_16K to listOf(
            FrameSize.FRAME_SIZE_160,
            FrameSize.FRAME_SIZE_320,
            FrameSize.FRAME_SIZE_480
        ),
        SampleRate.SAMPLE_RATE_32K to listOf(
            FrameSize.FRAME_SIZE_320,
            FrameSize.FRAME_SIZE_640,
            FrameSize.FRAME_SIZE_960
        ),
        SampleRate.SAMPLE_RATE_48K to listOf(
            FrameSize.FRAME_SIZE_480,
            FrameSize.FRAME_SIZE_960,
            FrameSize.FRAME_SIZE_1440
        )
    )

    /**
     * <p>
     * Valid Sample Rates and Frame Sizes for Silero VAD DNN model.
     * </p>
     */
    private val SILERO_VALID_FRAMES: Map<SampleRate, List<FrameSize>> = mapOf(
        SampleRate.SAMPLE_RATE_8K to listOf(
            FrameSize.FRAME_SIZE_256,
            FrameSize.FRAME_SIZE_512,
            FrameSize.FRAME_SIZE_768
        ),
        SampleRate.SAMPLE_RATE_16K to listOf(
            FrameSize.FRAME_SIZE_512,
            FrameSize.FRAME_SIZE_1024,
            FrameSize.FRAME_SIZE_1536
        )
    )

    /**
     * <p>
     * Validates the sample rate for a given VAD model.
     * </p>
     * @param model The VAD model to validate.
     * @param sampleRate The sample rate to validate.
     * @throws IllegalArgumentException if the sample rate is not supported by the model.
     */
    @JvmStatic
    fun validateSampleRate(model: Model, sampleRate: SampleRate) {
        require(getValidSampleRates(model).contains(sampleRate)) {
            "${model.name} doesn't support Sample Rate:${sampleRate.value}!"
        }
    }

    /**
     * <p>
     * Validates the frame size for a given VAD model and sample rate.
     * </p>
     * @param model The VAD model to validate.
     * @param sampleRate The sample rate to validate.
     * @param frameSize The frame size to validate.
     * @throws IllegalArgumentException if the frame size is not supported by the
     * model and sample rate combination.
     */
    @JvmStatic
    fun validateFrameSize(model: Model, sampleRate: SampleRate, frameSize: FrameSize) {
        require(getValidFrameSizes(model, sampleRate).contains(frameSize)) {
            "${model.name} doesn't support Sample rate:${sampleRate.value} and Frame Size:${frameSize.value}!"
        }
    }

    /**
     * <p>
     * Returns the valid frame sizes for a given VAD model and sample rate.
     * </p>
     * @param model The VAD model to retrieve valid frame sizes for.
     * @param sampleRate The sample rate to retrieve valid frame sizes for.
     * @return The list of valid frame sizes.
     */
    @JvmStatic
    fun getValidFrameSizes(model: Model, sampleRate: SampleRate): List<FrameSize> {
        return when (model) {
            Model.WEB_RTC_GMM -> WEB_RTC_VALID_FRAMES[sampleRate] ?: emptyList()
            Model.SILERO_DNN -> SILERO_VALID_FRAMES[sampleRate] ?: emptyList()
        }
    }

    /**
     * <p>
     * Returns the valid sample rates for a given VAD model.
     * </p>
     * @param model The VAD model to retrieve valid sample rates for.
     * @return The list of valid sample rates.
     */
    @JvmStatic
    fun getValidSampleRates(model: Model): List<SampleRate> {
        return when (model) {
            Model.WEB_RTC_GMM -> WEB_RTC_VALID_FRAMES.keys.toList()
            Model.SILERO_DNN -> SILERO_VALID_FRAMES.keys.toList()
        }
    }
}